File: 剑指offer/C++/PrintListFromTailToHead.cpp
#include <iostream>
#include <vector>

using namespace std;

struct ListNode
{
    int val;
    struct ListNode *next;
    ListNode(int x) : val(x), next(NULL)
    {
    }
};

/*
 * 遍历,从头扫到尾,逐个装入vector中,最后反转vector
 */
class Solution
{
public:
    vector<int> printListFromTailToHead(ListNode *head)
    {
        vector<int> res;
        while (head != NULL)
        {
            res.push_back(head->val);
            head = head->next;
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

int main(int argc, char **argv, char **envp)
{
    int n, val;
    while (cin >> n)
    {
        ListNode *head = new ListNode(0);
        ListNode *p = head;
        while (n--)
        {
            cin >> val;
            head->next = new ListNode(val);
            head = head->next;
        }
        vector<int> res = Solution().printListFromTailToHead(p->next);
        for (vector<int>::iterator iter = res.begin(); iter != res.end(); iter++)
        {
            cout << (*iter) << endl;
        }
    }
    return 0;
}

File: 剑指offer/Java/PrintListFromTailToHead.java
import java.util.ArrayList;
import java.util.Collections;

class ListNode {
    int val;
    ListNode next = null;
    
    ListNode(int val) {
        this.val = val;
    }
}


class Solution {
    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
        ArrayList<Integer> res = new ArrayList<>();
        while (listNode != null){
            res.add(listNode.val);
            listNode = listNode.next;
        }
        Collections.reverse(res);
        return res;
    }
}

public class PrintListFromTailToHead{
    public static void main(String[] args) {
        
    }
}